src/wallet/coinselection.cpp

```diff
@@ -246,6 +246,80 @@ util::Result<SelectionResult> SelectCoinsSRD(const std::vector<OutputGroup>& utx
     return max_tx_weight_exceeded ? ErrorMaxWeightExceeded() : util::Error();
 }
 
+util::Result<SelectionResult> SelectCoinsGG(std::vector<OutputGroup>& utxo_pool, CAmount target_value, CAmount change_fee, FastRandomContext& rng,
+                                             int max_weight)
+{
+    SelectionResult result(target_value, SelectionAlgorithm::GG);
+    std::priority_queue<OutputGroup, std::vector<OutputGroup>, MinOutputGroupComparator> heap;
+
+    // Include change for Gutter Guard Selector as we want to avoid making really small change if the selection just
+    // barely meets the target.
+    target_value += CHANGE_LOWER + change_fee;
+
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+    CAmount selected_lf_amount = 0;
+    size_t lf_group_count = 0;
+
+    for (const OutputGroup& group : utxo_pool) {
+        selected_lf_amount += group.GetSelectionAmount();
+        lf_group_count++;
+        if (selected_lf_amount >= target_value) {
+            break;
+        }
+    }
+
+    size_t gg_group_limit = lf_group_count + 3;
+
+    std::vector<size_t> indexes;
+    indexes.resize(utxo_pool.size());
+    std::iota(indexes.begin(), indexes.end(), 0);
+    Shuffle(indexes.begin(), indexes.end(), rng);
+
+    CAmount selected_eff_value = 0;
+    int weight = 0;
+    bool max_tx_weight_exceeded = false;
+    for (const size_t i : indexes) {
+        const OutputGroup& group = utxo_pool.at(i);
+        Assume(group.GetSelectionAmount() > 0);
+
+        // Add group to selection
+        heap.push(group);
+        selected_eff_value += group.GetSelectionAmount();
+        weight += group.m_weight;
+
+        // If the selection weight exceeds the maximum allowed size, remove the least valuable inputs until we
+        // are below max weight.
+        if (weight > max_weight) {
+            max_tx_weight_exceeded = true; // mark it in case we don't find any useful result.
+            do {
+                const OutputGroup& to_remove_group = heap.top();
+                selected_eff_value -= to_remove_group.GetSelectionAmount();
+                weight -= to_remove_group.m_weight;
+                heap.pop();
+            } while (!heap.empty() && weight > max_weight);
+        }
+
+        // If we have selected more output groups than permitted, drop OG with lowest effective value
+        while (!heap.empty() && heap.size() > gg_group_limit) {
+            const OutputGroup& to_remove_group = heap.top();
+            selected_eff_value -= to_remove_group.GetSelectionAmount();
+            weight -= to_remove_group.m_weight;
+            heap.pop();
+        }
```

Comment:
In 1e19f925bb3d2b74933d8cb6d3628477cdd58c32 "Add Gutter Guard Selector"

These two loops could probably be consolidated into a single one since they do the same thing, just under different conditions.

Reply:
I combined the two loops